### ISSUE TITLE
Change the internal IP

### ISSUE THREAD

**Issue by ArseniiPetrovich** [canonical/microk8s#2402]
Can I somehow change the IP that the node is using as internal IP? My nodes has an VPN interconnection, and it seems they communicate well through the VPN. However when it comes to setting up addons like DNS I'm having a "No route to host" issue. I tracked it down to microk8s trying to use the "Internal IPs" of nodes instead of using their VPN IPs.

**Comment by ktsakalozos** [comment]
@ArseniiPetrovich could you provide an example with IPs in your setup?

Normally the pods get an IP and the CNI and kube-proxy makes sure the communication is established among them. At which point do you get the "No route to host" error?

**Comment by ArseniiPetrovich** [comment]
https://drive.google.com/drive/folders/18GFyEQ9gf6seLBfM72rUbdBQGuUQ-PXp?usp=sharing

Here, 3 tarballs from all our nodes. Would that answer your question?
I want clusters to connect with `10.11.11.x` addresses, while they connect with their primary addresses, either internal or external ones.

**Comment by ArseniiPetrovich** [comment]
So I get an error when doing `mk8s enable/disable dns`

**Comment by ArseniiPetrovich** [comment]
FYI, one of the nodes is currently out of cluster, I'm testing different options to fix this, but this does not work with 2 nodes either.

**Comment by afeiszli** [comment]
I just ran into the same issue attempting the same process. 
1. Installed private interfaces on all Ubuntu servers using 10.99.x.x addresses (using Netmaker though this shouldn't matter)
2. Installed MicroK8s on top
3. Used the add-node command and selected the private interface
4. Ran the join command on the next node
5. repeat steps 3-4 on all nodes
6. Install CoreDNS

At this point it fails trying to reach the first node. Running "microk8s kubectl get nodes -o wide" I see that all nodes are using default interface IP's as their "internal IP" and no "external IP" is set.

**Comment by afeiszli** [comment]
Update: to set internal IP's you can run the following (run as root on each node):

echo --node-ip=$(ip address show dev YOUR_PRIVATE_INTERFACE | grep 'inet ' | awk -F ' ' '{print $2}' | sed 's/\/24//g') >> /var/snap/microk8s/current/args/kubelet; microk8s stop; microk8s start

Replacing YOUR_PRIVATE_INTERFACE with the private interface name. Idk if this is correct or causes any different issues, but it at least sets the internal IP correctly.

**Comment by ArseniiPetrovich** [comment]
@ktsakalozos Is there a valid way to do this?
I attempted the following to solve my issue:
1) Hardcode the IP in the Calico config (/var/snap/microk8s/current/args/cni-network/cni.yaml)
2) Hardcode the IP in the API-server config by setting `--advertise-address` in the /var/snap/microk8s/current/args/kube-apiserver
3) Hardcode the IP in the Kubelet config by setting `--node-ip` in the /var/snap/microk8s/current/args/kubelet

It worked when I started the node, but when I tried to join these nodes together it went completely nowhere with the following error logs:
```
Jul  6 23:46:39 one microk8s.daemon-kubelite[109803]: E0706 23:46:39.564207  109803 available_controller.go:508] v1beta1.metrics.k8s.io failed with: failing or missing response from https://10.152.183.165:443/apis/metrics.k8s.io/v1beta1: Get "https://10.152.183.165:443/apis/metrics.k8s.io/v1beta1": dial tcp 10.152.183.165:443: connect: no route to host
Jul  6 23:46:39 one microk8s.daemon-kubelite[109803]: E0706 23:46:39.633782  109803 kubelet.go:2291] "Error getting node" err="node \"one.us.vpc\" not found"
Jul  6 23:46:39 one microk8s.daemon-kubelite[109803]: E0706 23:46:39.734625  109803 kubelet.go:2291] "Error getting node" err="node \"one.us.vpc\" not found"
Jul  6 23:46:39 one microk8s.daemon-kubelite[109803]: E0706 23:46:39.834942  109803 kubelet.go:2291] "Error getting node" err="node \"one.us.vpc\" not found"
Jul  6 23:46:39 one microk8s.daemon-kubelite[109803]: E0706 23:46:39.936713  109803 kubelet.go:2291] "Error getting node" err="node \"one.us.vpc\" not found"
Jul  6 23:46:40 one microk8s.daemon-kubelite[109803]: E0706 23:46:40.037661  109803 kubelet.go:2291] "Error getting node" err="node \"one.us.vpc\" not found"
Jul  6 23:46:40 one microk8s.daemon-kubelite[109803]: E0706 23:46:40.138446  109803 kubelet.go:2291] "Error getting node" err="node \"one.us.vpc\" not found"
Jul  6 23:46:40 one microk8s.daemon-kubelite[109803]: E0706 23:46:40.239243  109803 kubelet.go:2291] "Error getting node" err="node \"one.us.vpc\" not found"
Jul  6 23:46:40 one microk8s.daemon-kubelite[109803]: E0706 23:46:40.339734  109803 kubelet.go:2291] "Error getting node" err="node \"one.us.vpc\" not found"
Jul  6 23:46:40 one microk8s.daemon-kubelite[109803]: E0706 23:46:40.440353  109803 kubelet.go:2291] "Error getting node" err="node \"one.us.vpc\" not found"
Jul  6 23:46:40 one microk8s.daemon-kubelite[109803]: E0706 23:46:40.541775  109803 kubelet.go:2291] "Error getting node" err="node \"one.us.vpc\" not found"
Jul  6 23:46:40 one microk8s.daemon-kubelite[109803]: E0706 23:46:40.542045  109803 reflector.go:138] k8s.io/client-go/informers/factory.go:134: Failed to watch *v1.Service: failed to list *v1.Service: Get "https://127.0.0.1:16443/api/v1/services?limit=500&resourceVersion=0": x509: certificate signed by unknown authority (possibly because of "crypto/rsa: verification error" while trying to verify candidate authority certificate "10.152.183.1")
```

**Comment by ArseniiPetrovich** [comment]
@afeiszli check my answer below. It works for one node, but when you try to join these nodes together - it breakes the whole cluster.

**Comment by ArseniiPetrovich** [comment]
```
Jul  6 23:59:59 one microk8s.daemon-kubelite[134568]: E0706 23:59:59.270000  134568 kubelet_node_status.go:586] "Failed to set some node status fields" err="failed to validate nodeIP: node IP: \"10.11.11.5\" not found in the host's network interfaces" node="one.us.vpc"
Jul  7 00:00:00 one microk8s.daemon-kubelite[134568]: E0707 00:00:00.876002  134568 available_controller.go:508] v1beta1.metrics.k8s.io failed with: failing or missing response from https://10.152.183.165:443/apis/metrics.k8s.io/v1beta1: Get "https://10.152.183.165:443/apis/metrics.k8s.io/v1beta1": dial tcp 10.152.183.165:443: connect: no route to host

```
It actually worked on the second try, but still put some weird logs...

**Comment by ArseniiPetrovich** [comment]
Realized that for some reason after joining cluster `--node-ip` in the Kubelet file of joining node changes to the IP of the node where it joins.
For example:
node 1 (ip 1): microk8s add-node
node 2 (ip 2): microk8s join ...
node 2 get the ip1 for in kubelet config

**Comment by ArseniiPetrovich** [comment]
Okay, I were able to join all the nodes and then manually override IP addresses for them. However when trying to run some pods I'm getting the following error:
```

Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "b9b97afc5e9b38f8dd320c90e1fe96bb6268ef46df376ff694730718deea03b2": error getting ClusterInformation: Get https://[10.152.183.1]:443/apis/crd.projectcalico.org/v1/clusterinformations/default: dial tcp 10.152.183.1:443: connect: no route to host
```

**Comment by ArseniiPetrovich** [comment]
Flushed IP Tables as suggested in other issues. Still getting the error, but with timeout now:
```
Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "0c4180725ae059bb0031e08bdabbfda5b63effe07784f5522f235ede8286886f": error getting ClusterInformation: Get https://[10.152.183.1]:443/apis/crd.projectcalico.org/v1/clusterinformations/default: dial tcp 10.152.183.1:443: i/o timeout
```

**Comment by afeiszli** [comment]
That's strange. Have you checked with a simple ping to make sure nodes are still able to talk to each other over the private addresses (you can also run "wg show" and check the last handshake)? If they are not, it may be necessary to restart the network interface after an iptables flush depending on the rules that were put in place, which can be done with "netclient pull -n <networkname>" on each node. Also make sure ip forwarding is enabled.

**Comment by ktsakalozos** [comment]
Can you make sure (probably by editing the /etc/hosts on each node) that the hostname of each node resolves to the IP on the VPN?

**Comment by afeiszli** [comment]
This fixed the ClusterIP issue for me. For posterity, a quick hack to run on each node, that will take the all node hostnames and advertised addresses, and insert them into /etc/hosts: 

echo "$(microk8s kubectl get nodes -o wide | awk 'NR>1{ print $6, $1 }')" >> /etc/hosts

**Comment by ArseniiPetrovich** [comment]
@afeiszli Did this solved this timeout issue we discussed today?

**Comment by afeiszli** [comment]
@ArseniiPetrovich, @ktsakalozos mentioned that hostnames may need to be reachable before running the cluster install. I've reinstalled by first setting up hostnames. I still recieved a timeout, but it appears to be better at least for now.

**Comment by afeiszli** [comment]
I believe I've found a key piece of this. If you do a "kubectl describe svc kubernetes -n default", you will see the endpoints are still pointing to the public IP's. To fix this you must stop each node (microk8s stop), and edit "/var/snap/microk8s/current/args/kube-apiserver", adding in the following line: --advertise-address=<private address>

When the node comes back up, the endpoint should be correct. This fixed one of my remaining timeout errors. Will leave cluster up for a while to see if other errors are encountered but network seems to be behaving now.

**Comment by ArseniiPetrovich** [comment]
@afeiszli When I did this last time I received an error about certificates, when ned nodes were not able to join the Microk8s server. Can you try under this setup adding new node?

**Comment by ArseniiPetrovich** [comment]
@afeiszli @ktsakalozos Small friendly ping here :)

**Comment by dannyfranca** [comment]
Same problem here, incredible that so many kubernetes distributions have the same problem. I only could get it to work with K3S. Ir is so simples to provide some --node-ip flag or even respect the interface used on the first node to join.

**Comment by dannyfranca** [comment]
Just for the issue record, my use case is: I have VPSs in differente cloud providers and bare metals with residential IPs. So I don't  have an private network in common. To solve this, I'm using Wireguard to wrap all machines in same private network (interface wg0).

What I expect: every nodes joining through IP provided by wg0 interface having internal IP described in wg0.
What I got: the main Ubuntu interface's IP

Any way to make it happen?

I just need every nodes talk each other through a choosed interface or event that one used on join command.

**Comment by wisaaco** [comment]
Hello all,

I have the same issue here. My workers have two interfaces and microk8s.join gets the "secondary" one. I try several comments but not one works. Apparently, kubelet ignores the --node-ip flag, and the BGP IP detection method in cni-network/cni.yaml file.

In my case, the ubuntu1 is a simple worker.
```
~$ kubectl get nodes -o wide
NAME      STATUS   ROLES    AGE     VERSION                    INTERNAL-IP     EXTERNAL-IP   OS-IMAGE             KERNEL-VERSION     CONTAINER-RUNTIME
ubuntu1   Ready    <none>   15m     v1.23.1-2+b7120affd6631a   10.162.X.Y   <none>        Ubuntu 20.04.3 LTS   5.4.0-1050-raspi   containerd://1.5.7
dellemc   Ready    <none>   3h24m   v1.23.1-2+9e8d1bf4219080   172.30.248.86   <none>        Ubuntu 20.04.3 LTS   5.4.0-89-generic   containerd://1.5.7
```
and it has the following, routes:
```
# sudo ip route list
default via 169.254.66.80 dev eth0 proto static 
default via 10.162.16.1 dev wlan0 proto dhcp src 10.162.X.Y metric 600 
...
```

I follow this thread and the related ones but I could not find any solution. I appreciate some kind of help with respect to this configuration.
Best

**Comment by dannyfranca** [comment]
Oly k3s solves that problem. incredible how this simple use case is a huge problem across many frameworks.

**Comment by wisaaco** [comment]
Hi @neoaggelos ,

In one of the attempts, I tried to remove one default interface in a fresh installation but the snap.microk8s.daemon-kubelite service did not start.

In any case, I follow these steps:
- In the master: 

```
$ microk8s add-node
From the node you wish to join to this cluster, run the following:
microk8s join 172.30.248.86:25000/587970307c112959444429819caa0abc/e31414f43733

Use the '--worker' flag to join a node as a worker not running the control plane, eg:
microk8s join 172.30.248.86:25000/587970307c112959444429819caa0abc/e31414f43733 --worker

If the node you are adding is not reachable through the default interface you can use one of the following:
microk8s join 172.30.248.86:25000/587970307c112959444429819caa0abc/e31414f43733
microk8s join 172.17.0.1:25000/587970307c112959444429819caa0abc/e31414f43733
microk8s join 192.168.49.1:25000/587970307c112959444429819caa0abc/e31414f43733
microk8s join 172.18.0.1:25000/587970307c112959444429819caa0abc/e31414f43733
microk8s join fc00:f853:ccd:e793::1:25000/587970307c112959444429819caa0abc/e31414f43733
```
My master also has to interfaces:
```
2: eno1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 2c:ea:7f:e0:17:70 brd ff:ff:ff:ff:ff:ff
    inet 172.30.248.86/22 brd 172.30.251.255 scope global dynamic eno1
3: eno2: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 2c:ea:7f:e0:17:71 brd ff:ff:ff:ff:ff:ff
    inet 169.254.66.80/16 brd 169.254.255.255 scope global eno2
```
- And from the output,  I chose the next command to run in the worker:
``` microk8s join 172.30.248.86:25000/aa48a0e60504b58ab9a0d6191206f1ae/e31414f43733 --worker ```
If I choose this ```microk8s join 192.168.49. ...```, the worker connects to the master but it is not visible inside the master. 

Note all devices ping each other. In the master, eno1 provides internet access and eno2 is local. 
The workers are connected via eno2 to the master.

The content of  /var/snap/microk8s/current/var/kubernetes/backend/cluster.yaml is:
```
$ cat /var/snap/microk8s/current/var/kubernetes/backend/cluster.yaml
- Address: 172.30.248.86:19001
  ID: 3297041220608546238
  Role: 0
```
Thanks!

**Comment by ephillipe** [comment]
I'm having issues after deleting a second node from a HA cluster without "leaving" from cluster. It occurs because I'm using the Kubernetes autoscaler.

Behavior: After second node be removed, the master stuck forever.

Notes:
- I have two interfaces, one public and another internal (vpn)
- I tried remove the second node reference from /var/snap/microk8s/current/var/kubernetes/backend/*.yaml files manually
- I tried restart the master node
- My internal network (vpc) is working.

I had any success. 

Thanks an advance by the help.

[inspection-report-20220423_054452.tar.gz](https://github.com/canonical/microk8s/files/8546294/inspection-report-20220423_054452.tar.gz)

**Comment by usersina** [comment]
I found a way to monkey patch the network routes by basically changing the default interface. However **this should not be considered as a solution since it BREAKS SSH.**

Supposing this is the output of `route -n`
```
Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
0.0.0.0         10.255.255.1    0.0.0.0         UG    0      0        0 ens192
10.1.68.128     10.1.68.128     255.255.255.192 UG    0      0        0 vxlan.calico
10.152.183.0    192.168.1.1     255.255.255.0   UG    0      0        0 ens224
10.255.255.1    0.0.0.0         255.255.255.255 UH    0      0        0 ens192
192.168.1.0     0.0.0.0         255.255.255.0   U     0      0        0 ens224
```

```bash
# Remove the default WAN interface (ens192 in my case)
sudo route del -net 0.0.0.0 gw 10.255.255.1 netmask 0.0.0.0 dev ens192 => WILL BREAK SSH

# Change the default interface to redirect default traffic to LAN interface (ens224 in my case)
sudo ip route add default via 192.168.1.1 dev ens224
```

If you did this on a server and want to get SSH back, you have to run this from a KVM console. (I was there Gandalf)
```bash
# Set rule to redirect default traffic to ens192 WAN interface (Will restore SSH)
sudo ip route add default via 10.255.255.1 dev ens192
```

**Comment by usersina** [comment]
> > Realized that for some reason after joining cluster `--node-ip` in the Kubelet file of joining node changes to the IP of the node where it joins. For example: node 1 (ip 1): microk8s add-node node 2 (ip 2): microk8s join ... node 2 get the ip1 for in kubelet config
> 
> I'm also seeing this. I have to change `/var/snap/microk8s/current/args/kubelet` again **after** joining to make the changes stick. Obviously a bug!

I tried experimenting with the [`--address`](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/#options) option in `/var/snap/microk8s/current/args/kubelet` set to the address of the interface I want to use (192.168.1.0 for example) but it didn't work out. I also tried out [`--bind-address`](https://stackoverflow.com/a/66259599/10543130) instead of `--advertise-address` but no luck there too.

The reason I'm trying that out is that if you have dns enabled and you want to join a node without changing the node ip first, it always timeouts...

**Comment by usersina** [comment]
> > Realized that for some reason after joining cluster `--node-ip` in the Kubelet file of joining node changes to the IP of the node where it joins. For example: node 1 (ip 1): microk8s add-node node 2 (ip 2): microk8s join ... node 2 get the ip1 for in kubelet config
> 
> I'm also seeing this. I have to change `/var/snap/microk8s/current/args/kubelet` again **after** joining to make the changes stick. Obviously a bug!

I tried `chattr +i /var/snap/microk8s/current/args/kubelet` to make the file immutable before joining. If I then execute the join command, I get a **PermissionError**

```bash
ilg@dev-kube-worker-1:~$ microk8s join 192.168.1.1:25000/597b86fd8da7b0c6c9c7cfd8bd7668a8/e338bb68685f --worker

Contacting cluster at 192.168.1.1
Traceback (most recent call last):
  File "/snap/microk8s/3272/scripts/cluster/join.py", line 968, in <module>
    join(prog_name="microk8s join")
  File "/snap/microk8s/3272/usr/lib/python3/dist-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/snap/microk8s/3272/usr/lib/python3/dist-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/snap/microk8s/3272/usr/lib/python3/dist-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/snap/microk8s/3272/usr/lib/python3/dist-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/snap/microk8s/3272/scripts/cluster/join.py", line 961, in join
    join_dqlite(connection_parts, verify, worker)
  File "/snap/microk8s/3272/scripts/cluster/join.py", line 743, in join_dqlite
    join_dqlite_worker_node(info, master_ip, master_port, token)
  File "/snap/microk8s/3272/scripts/cluster/join.py", line 824, in join_dqlite_worker_node
    store_base_kubelet_args(info["kubelet_args"])
  File "/snap/microk8s/3272/scripts/cluster/join.py", line 548, in store_base_kubelet_args
    with open(args_file, "w") as fp:
PermissionError: [Errno 1] Operation not permitted: '/var/snap/microk8s/3272/args/kubelet'
```

Why is the file being overridden anyway? It doesn't make sense since it's an `args` file

**Comment by usersina** [comment]
> > Realized that for some reason after joining cluster `--node-ip` in the Kubelet file of joining node changes to the IP of the node where it joins. For example: node 1 (ip 1): microk8s add-node node 2 (ip 2): microk8s join ... node 2 get the ip1 for in kubelet config
> 
> I'm also seeing this. I have to change `/var/snap/microk8s/current/args/kubelet` again **after** joining to make the changes stick. Obviously a bug!

I found a "fix" for this if you can call it that. Let's suppose you want to create a cluster from these two nodes:
- master-node: `192.168.1.1`
- worker-node-1: `192.168.1.2` 

In **master-node**:
1. View the nodes in the cluster with `microk8s kubectl get nodes -o wide`
2. Set `--node-ip` in `/var/snap/microk8s/current/args/kubelet` to `192.168.1.2` (IP of **worker-node-1**, weird I know but DO NOT RESTART THE CLUSTER)
3. Execute the add command with `microk8s add-node`
4. Copy the join command.

In **worker-node-1**:
5. Execute the join command.

In **master-node**:
6. Set `--node-ip` in `/var/snap/microk8s/current/args/kubelet` back to the original `192.168.1.1`.

**Comment by stygmate** [comment]
Any solution for this bug ? with this one we cannot consider microk8s production ready.

**Comment by neoaggelos** [comment]
> Any solution for this bug ? with this one we cannot consider microk8s production ready.

We are currently in the process of improving the experience of configuring MicroK8s in environments with multiple interfaces, air-gap deployments, etc.

There is already https://github.com/canonical/microk8s/pull/3216, which fixes the issues that were discussed regarding the handling of the `--node-ip` kubelet argument. In the coming days, we will also be working to improve the documentation around this. We welcome any feedback around the current pain points, which will in turn make the documentation much more effective and focused on solving the issue at hand.

**Comment by devmcp** [comment]
> There is already https://github.com/canonical/microk8s/pull/3216, which fixes the issues that were discussed regarding the handling of the --node-ip kubelet argument...

So does that mean this should be fixed? Should I now be able to set an internal IP with `--node-id`? Where do I set this config?

I've tried this:

```
echo '--node-ip=192.168.62.63' | tee -a /var/snap/microk8s/current/args/kubelet
sudo systemctl restart snap.microk8s.daemon-kubelet.service
```

which is what the [docs](https://microk8s.io/docs/configuring-services) seems to suggest, but this IP address is not being reflected in `microk8s.kubectl get nodes -o wide`

**Comment by neoaggelos** [comment]
For all interested in this, see [Configure Host Interfaces used by MicroK8s](https://discuss.kubernetes.io/t/configure-host-interfaces-used-by-microk8s/21508). Any feedback is appreciated!

**Comment by neoaggelos** [comment]
@devmcp apologies for missing this. If it is still relevant, you should do a `sudo snap restart microk8s`

For more context, `sudo snap restart microk8s.daemon-kubelet` is only valid for pre-1.19 installs. Starting from MicroK8s 1.19, it should be `sudo snap restart microk8s.daemon-kubelite`. This is because starting from 1.19,  `kubelet` (and all the other kubernetes components) are running under a single process (called kubelite) as an optimization for reduced memory and CPU footprint on the host.

As a simpler solution that works everywhere `sudo snap restart microk8s` will do the right thing and you don't have to know about any of this.

**Comment by balboah** [comment]
As an alternative for hard coding an IP in cni and/or setting `--node-ip`, you can specify an interface for calico to fetch the IPs from:

```yaml
            - name: IP
               value: "autodetect"
            - name: IP_AUTODETECTION_METHOD
               value: interface=foo
            - name: IP6
               value: autodetect
            - name: IP6_AUTODETECTION_METHOD
               value: interface=foo
```

At least for me, this will set the correct internal IP.

However, as I replied in the article referenced by @neoaggelos:
setting `--advertise-address` (to also get the correct IPs in the kubelet service and --worker nodes) can only be done **after** a node already has joined or it will break.

**Comment by neoaggelos** [comment]
> setting --advertise-address (to also get the correct IPs in the kubelet service and --worker nodes) can only be done after a node already has joined or it will break.

This is in fact correct, thanks for pointing it out, will update the documentation accordingly.

I just want to make sure that it is clear that Kubernetes may use multiple interfaces, this is how to change each one:

- `--advertise-address` in the `kube-apiserver` arguments, configures the server address that is used to reach the API server by clients, and can be found in `kubectl get endpoints kubernetes`
- `--node-ip` in the `kubelet` arguments, configures the internal IP address used by the kubelet. this is the one shown in `kubectl get node $node -o wide`
- Calico configuration is orthogonal to the above, and is only used to configure which underlay interface is used by the overlay CNI network for pod-to-pod communication in the cluster.

Hope this helps!

**Comment by usersina** [comment]
Alright, so I'm building on the previous answers to document what exactly worked for me and to save my future self some headaches.

Let's say I have two VMs, both VMs have a public interface and a private LAN interface with the public one being the default. That's a problem, since joining will usually use the default public interface but I want the join to be done using the private LAN network. I also want the inter-node communication to be done using the private LAN interface address, so I have to set that up as well. (Ultimately `--node-ip` and `--advertise-address`)

Supposing that the private LAN address is `192.168.1.0/24` with:

- `kube-controller` at `192.168.1.1`
- `worker-node` at `192.168.1.2`

## 1. Cleaning leftover configuration
I had already done multiple join attempts from `worker-node`, using different settings every time but nothing worked in the end. The reason for that is most likely some leftover cni configuration on the `worker-node` itself. (You can see that from the cni Pod logs on the `worker-node`)

To completely remove that, do the following:

- Leave the cluster if not already
```
(worker-node) $ microk8s leave
```

- Only after the leave command finishes, remove the node from the controller
```
(kube-controller) $ microk8s remove-node worker-node
```

If you get a permission denied error, run the command again. 

**Important:** You might also want to kill all pods that were previously running on `worker-node` since it's not guaranteed that they will be killed when the leave command is issued. (I use k9s thanks to how convenient it is, but you could use a fancy `kubectl delete --force` command that targets the hostname `worker-node`)

- Completely reset microk8s on `worker-node`
```
(worker-node) $ sudo snap remove --purge microk8s
(worker-node) $ sudo snap install microk8s --classic
```

- Make sure that the node did indeed leave the cluster and that it has no leftover `Pending` or `Terminating` pods left behind.
```
(kube-controller) $ kubectl get all --all-namespaces
(kube-controller) $ kubectl get nodes -o wide
```

You should now safely attempt to re-join the cluster with the next step. Note that you don't have to reset the `kube-controller` microk8s instance.

## 2. Joining on the LAN interface

I recommend temporarily disabling the ingress add-on if it's enabled so that we can test the pod scheduling with it.
```
(kube-controller) $ microk8s disable ingress
```

- Configure the `--node-ip` and `--advertise-address` on the controller (note the used IP address)
```bash
(kube-controller) $ sudo echo --node-ip=192.168.1.1 >> /var/snap/microk8s/current/args/kubelet
(kube-controller) $ sudo echo --advertise-address=192.168.1.1 >> /var/snap/microk8s/current/args/kube-apiserver
```

- Create the join command
```bash
(kube-controller) $ microk8s add-node

(omitted)
If the node you are adding is not reachable through the default interface you can use one of the following:
microk8s join 192.168.1.1:25000/abababababab/ababab
(omitted)
```

- Run the join command and don't forget the `--worker` flag
```bash
(worker-node) $ microk8s join 192.168.1.1:25000/abababababab/ababab --worker
```

- Enable ingress to schedule a Pod for each node
```
(kube-controller) $ microk8s enable ingress
```

- Check if an ingress Pod is running for the `worker-node`
```bash
(kube-controller) $ kubectl get pods --namespace ingress -o wide
```

The expected output is the following, if the status is still `Pending` you might want to check the logs of that pod.
```
NAME                                      READY   STATUS    RESTARTS   AGE   IP            NODE           
nginx-ingress-microk8s-controller-qt8v7   1/1     Running   0          49m   10.1.68.163   kube-controller
nginx-ingress-microk8s-controller-2kzgr   1/1     Running   0          49m   10.1.216.1    worker-node   
```

### Useful information
I already setup hostname and local hostnames resolution as shown below.

```bash
(kube-controller) $ cat /etc/hostname
kube-controller

(kube-controller) $ cat /etc/hosts
    (omitted)
    127.0.1.1 localhost kube-controller
    127.0.0.1 localhost
    
    192.168.1.2 worker-node
    (omitted)
```

```bash
(worker-node) $ cat /etc/hostname
worker-node

(worker-node) $ cat /etc/hosts
    (omitted)
    127.0.1.1 localhost worker-node
    127.0.0.1 localhost
    
    192.168.1.1 kube-controller
    (omitted)
```

These are the versions of Ubuntu and microk8s that I used
```bash
(both nodes) $ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 22.04.2 LTS
Release:	22.04
Codename:	jammy

(both-nodes) $ snap info microk8s
(omitted)
1.26/stable:           v1.26.1         2023-02-07 (4595) 176MB classic
(omitted)
```

### Related issues:
- https://github.com/canonical/microk8s/issues/3101
- https://github.com/canonical/microk8s/issues/3149
- https://github.com/canonical/microk8s/pull/3216

**Comment by pishangujeniya** [comment]
@usersina I have one master cluster node from one of the cloud service provider and I am trying to add Azure VM as a worker node to my cluster, via public IPs.

After adding node 
following is the output of `kubectl get nodes -o wide`

```
NAME                           STATUS   ROLES    AGE   VERSION   INTERNAL-IP   EXTERNAL-IP   OS-IMAGE       KERNEL-VERSION     CONTAINER-RUNTIME
worker-gpu-dev               Ready    <none>   33m   v1.29.2   192.168.0.4   <none>        Ubuntu 23.10   6.5.0-1019-azure   containerd://1.6.28
master.contaboserver.net   Ready    <none>   28d   v1.29.2   5.182.17.85   <none>        Ubuntu 23.10   6.5.0-26-generic   containerd://1.6.28
```

I have failure in pod to pod communication across the nodes, and I guess the reason is `worker-gpu-dev` node is trying to use `192.168.0.4` which is the Azure internal virtual network private IP assigned to the node, and also it has public IP too.

To solve this, I am trying to make sure that the worker node uses public IP just like how the master node is having its own public IP.

I added node public IP the file before joining cluster, and after installing microk8s

`sudo cat /var/snap/microk8s/current/args/kubelet`

```
--resolv-conf=/run/systemd/resolve/resolv.conf
--kubeconfig=${SNAP_DATA}/credentials/kubelet.config
--cert-dir=${SNAP_DATA}/certs
--client-ca-file=${SNAP_DATA}/certs/ca.crt
--anonymous-auth=false
--root-dir=${SNAP_COMMON}/var/lib/kubelet
--fail-swap-on=false
--eviction-hard="memory.available<100Mi,nodefs.available<1Gi,imagefs.available<1Gi"
--container-runtime-endpoint=${SNAP_COMMON}/run/containerd.sock
--containerd=${SNAP_COMMON}/run/containerd.sock
--node-labels="microk8s.io/cluster=true,node.kubernetes.io/microk8s-controlplane=microk8s-controlplane"
--authentication-token-webhook=true
--read-only-port=0
--tls-cipher-suites=TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_RSA_WITH_AES_256_GCM_SHA384,TLS_RSA_WITH_AES_128_GCM_SHA256
--serialize-image-pulls=false
--node-ip=52.170.129.4
```

Now after `sudo microk8s join 5.182.17.85:25000/long_string/short_string --worker`

The node ip from the file gets removed automatically.

`sudo cat /var/snap/microk8s/current/args/kubelet`

```
--cluster-dns=10.152.183.10
--cluster-domain=cluster.local
--resolv-conf=/run/systemd/resolve/resolv.conf
--kubeconfig=${SNAP_DATA}/credentials/kubelet.config
--cert-dir=${SNAP_DATA}/certs
--client-ca-file=${SNAP_DATA}/certs/ca.remote.crt
--anonymous-auth=false
--root-dir=${SNAP_COMMON}/var/lib/kubelet
--fail-swap-on=false
--feature-gates=
--eviction-hard="memory.available<100Mi,nodefs.available<1Gi,imagefs.available<1Gi"
--container-runtime-endpoint=${SNAP_COMMON}/run/containerd.sock
--containerd=${SNAP_COMMON}/run/containerd.sock
--node-labels=microk8s.io/cluster=true,node.kubernetes.io/microk8s-worker=microk8s-worker
--authentication-token-webhook=true
--read-only-port=0
```

Can you help me out, why my internal IP still shows the private IP which is totally no accessible between the nodes.
I have kept all the ports open between the two machines.

**Comment by Dzeri96** [comment]
It's almost 2025 and this issue has officially not been resolved, only closed by the bot for an unknown reason. @ktsakalozos since the assigned person does not seem to work for canonical anymore, would it make sense to reopen this issue and assign it to an active maintainer?

**Comment by mmraz** [comment]
> echo --node-ip=$(ip address show dev YOUR_PRIVATE_INTERFACE | grep 'inet ' | awk -F ' ' '{print $2}' | sed "s//24//g") >> /var/snap/microk8s/current/args/kubelet; microk8s stop; microk8s start

This only works if your CIDR block for the subnet is a `/24`.  I run with pod IPs being assigned and 254 usable addresses isn't enough.  We use a `/16` and I was wondering why this command was failing.  I adapted it to use a regex so that you don't need to know the block size prior to execution and I can add this to the user data script for ccloud-init first boot execution.  Our boot image already has microk8s installed.

```
echo --node-ip=$(ip address show dev YOUR_PRIVATE_INTERFACE | grep 'inet ' | awk -F ' ' '{print $2}' | sed "s/\/[0-9]\+$//g") >> /var/snap/microk8s/current/args/kubelet; microk8s stop; microk8s start
```
